api/src/test/java/org/pmiops/workbench/workspaces/WorkspacesControllerTest.java

```diff
@@ -787,22 +793,156 @@ public void testUpdateWorkspace() throws Exception {
   }
 
   @Test
-  public void testUpdateWorkspace_freeTierBilling() throws Exception {
+  public void testUpdateWorkspace_freeTierBilling_usesCorrectProvider() throws Exception {
     Workspace ws = createWorkspace();
     ws = workspacesController.createWorkspace(ws).getBody();
 
-    verify(endUserCloudbillingProvider.get()).projects();
-    verifyZeroInteractions(serviceAccountCloudbillingProvider.get());
+    // Creating the workspace with a user provided billing account
+    endUserCloudbilling = TestMockFactory.createMockedCloudbilling();
+    serviceAccountCloudbilling = TestMockFactory.createMockedCloudbilling();
 
     UpdateWorkspaceRequest request = new UpdateWorkspaceRequest();
     ws.setBillingAccountName(workbenchConfig.billing.freeTierBillingAccountName());
     request.setWorkspace(ws);
     workspacesController.updateWorkspace(ws.getNamespace(), ws.getId(), request);
 
-    verifyNoMoreInteractions(endUserCloudbillingProvider.get());
+    verifyZeroInteractions(endUserCloudbillingProvider.get());
     verify(serviceAccountCloudbillingProvider.get()).projects();
   }
 
+  @Test
+  public void testUpdateWorkspace_freeTierBilling_noCreditsRemaining() throws Exception {
+    Workspace ws = createWorkspace();
```

Comment:
[opt] I tend to try to avoid abbreviations for var names in Java, though they're prevalent enough in our codebase I wouldn't block on this.